include/maths/CBoostedTreeImpl.h

```diff
@@ -101,14 +102,13 @@ class MATHS_EXPORT CBoostedTreeImpl final {
     std::size_t columnHoldingDependentVariable() const;
 
     //! Get the number of columns training the model will add to the data frame.
-    constexpr static std::size_t numberExtraColumnsForTrain() {
+    static std::size_t numberExtraColumnsForTrain(std::size_t numberLossParameters) {
         // We store as follows:
-        //   1. The predicted value for the dependent variable
+        //   1. The predicted values for the dependent variables
         //   2. The gradient of the loss function
-        //   3. The curvature of the loss function
+        //   3. The upper triangle of the hessian of the loss function
         //   4. The example's weight
-        // In the last four rows of the data frame.
-        return 4;
+        return numberLossParameters * (numberLossParameters + 5) / 2 + 1;
```

Comment:
nit: where does 5 come from? In Line 599 it's just `numberLossParameters * (numberLossParameters + 1) / 2`.

Reply:
This is `prediction + gradient + curvature = numberLossParameters + numberLossParameters + numberLossParameters * (numberLossParameters + 1) / 2 = numberLossParameters * (numberLossParameters + 1 + 4) / 2`